File: JSONPlaces/NearbyPlacesViewController.m
//
//  NearbyPlacesViewController.m
//  JSONPlaces
//
//  Created by Adrian Kubała on 12.02.2017.
//  Copyright © 2017 Adrian. All rights reserved.
//

#import "NearbyPlacesViewController.h"
#import "UITableViewController+Gestures.h"
#import "PlaceView.h"
#import "Place.h"
#import "PlaceViewController.h"

@implementation NearbyPlacesViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  
  [self addLongPressGestureRecognizerWithSelector:@selector(showPlaceDetails:)];
}

# pragma mark UITableViewDataSource

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
  return self.nearbyPlaces.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
  PlaceView *cell = (PlaceView *) [self.tableView dequeueReusableCellWithIdentifier:@"NearbyPlaceView" forIndexPath:indexPath];
  
  Place *place = self.nearbyPlaces[indexPath.row];
  cell.pinImageView.image = place.pinImage;
  cell.label.text = place.name;
  
  return cell;
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
  PlaceViewController *placeViewController = (PlaceViewController *) segue.destinationViewController;
  UITableViewCell *cell = (UITableViewCell *) sender;
  long row = [self.tableView indexPathForCell:cell].row;
  placeViewController.place = self.nearbyPlaces[row];
}

- (void)showPlaceDetails:(UIGestureRecognizer *)recognizer {
  if (recognizer.state == UIGestureRecognizerStateBegan) {
    CGPoint pressLocation = [recognizer locationInView:self.tableView];
    NSIndexPath *pressedIndexPath = [self.tableView indexPathForRowAtPoint:pressLocation];
    
    Place *place = self.nearbyPlaces[pressedIndexPath.row];
    CLLocation *location = place.location;
    CLLocationDegrees latitude = location.coordinate.latitude;
    CLLocationDegrees longitude = location.coordinate.longitude;
    double distance = place.distance;
    
    NSString *message = [NSString stringWithFormat:@"Szerokość: %f\nDługość: %f\nOdległość: %.2f km", latitude, longitude, distance];
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Współrzędne" message:message preferredStyle:UIAlertControllerStyleAlert];
    
    UIAlertAction *action = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
    [alert addAction:action];
    
    [self presentViewController:alert animated:true completion:nil];
  }
}

@end

File: JSONPlaces/NearbyPlacesViewController.swift
//
//  NearbyPlacesViewController.swift
//  JSONPlaces
//
//  Created by Adrian on 11.08.2016.
//  Copyright © 2016 Adrian. All rights reserved.
//

import UIKit

class NearbyPlacesViewController: UITableViewController {
  var nearbyPlaces: [Place] = []
  
  override func viewDidLoad() {
    addLongPressGestureRecognizer(with: #selector(showPlaceDetails(recognizer:)))
  }
  
  override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    return nearbyPlaces.count
  }
  
  override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    let cell = tableView.dequeueReusableCell(withIdentifier: "NearbyPlaceView", for: indexPath) as! PlaceView
    
    cell.pinImageView.image = nearbyPlaces[indexPath.row].pinImage
    cell.label.text = nearbyPlaces[indexPath.row].name
    
    return cell
  }
  
  override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
    let placeViewController = segue.destination as! PlaceViewController
    let cell = sender as! UITableViewCell
    let row = tableView.indexPath(for: cell)!.row
    placeViewController.place = nearbyPlaces[row]
  }
  
  func showPlaceDetails(recognizer: UIGestureRecognizer) {
    if recognizer.state == UIGestureRecognizerState.began {
      let swipeLocation = recognizer.location(in: tableView)
      if let swipedIndexPath = tableView.indexPathForRow(at: swipeLocation) {
        let location = nearbyPlaces[swipedIndexPath.row].location
        let latitude = location?.coordinate.latitude
        let longitude = location?.coordinate.longitude
        let distance = nearbyPlaces[swipedIndexPath.row].distance
        
        let alert = UIAlertController(title: "Współrzędne",
                                      message: "Szerokość: \(latitude)\nDługość: \(longitude)\nOdległość: \(String(format:"%.2f", distance)) km",
          preferredStyle: UIAlertControllerStyle.alert)
        alert.addAction(UIAlertAction(title: "OK", style: UIAlertActionStyle.default, handler: nil))
        present(alert, animated: true, completion: nil)
      }
    }
  }
}
